Compile time path

thehiddenwaffle opened the pull request:
Here's the big one, lots of design choices to be made. I tried to keep everything as simple as possible for now but in the future I could definitely remove a couple hundred lines(ToTokens  impls mostly) with declarative macros and once it stabilizes and has a full test suite I can work on better error reporting from the macro side and the FromPest side

I have tried to list out all the options that I have thought of, as well as the one that I think is best but I'm certainly not going to make those choices without presenting them to you and ensuring you agree. These are just my suggestions and I'm happy to expand on my thought process but ultimately it's your choice of course.

In this PR I have not integrated anything I created with the main functionality of this repository because there are still design choices to make, because its a 2600 line addition, and because it's barely tested.

### Summary:
- Added a new crate with a common ast called `jsonpath-ast`
  - Common structures shared between syn and pest, allowing them to parse into the same things
  - All syn related implementations are locked behind a crate feature "compiled-path", ensuring that users who do not need compiled paths via the `json_path!` macro will not suffer longer build times
    - Work could still be done to completely eliminate syn and other macro related dependencies when "compiled-path" is toggled off, currently not possible
  - Includes some helper types(`PestIgnoredPunctuated`, `PestLiteralWithoutRule`) for when pest checks that punctuation exist but doesn't parse it, so syn still needs to check that it exists, and FromPest needs to ignore it(because it already guaranteed it) during pest::Rules -> AST conversion
  - Current ToTokens implementation has each struct generate itself as tokens, ie: `Main` writes out "Main::new(....inners)"
    - Could be changed to be anything at all(see Macro Output below)
- Added new proc macro crate jsonpath-rust-impl
  - Provides the `json_path!` proc macro, depends on jsonpath-ast with feature "compiled-path" enabled
  - Has a test suite I generated using the basic.json from the rfc and trybuild. I think this would be a good approach, but I'm open to better ideas and therefore only did basic.json not the whole RFC
    - Even with trybuild there are still choices, see Testing below
  - Not everything can be parsed by syn at compile time, see Macro Limitations below


### Macro Limitations:
- single quote strings break the macro parser because in rust only chars are single quoted
    - rust has `r"# #"` strings so in my opinion this is no impact, we can provide a sed regex?
- due to constraints rust has put on it's identifiers, emoji in member_name_shorthand is not allowed(works in both string literals and bracket field access`json_query!( $["☺"] )`)
- we could solve these if we changed the input to a string literal, *however*:
    - this would make error messages and syntax highlighting severely more complex to implement, and currently syn is giving them to us absolutely free(in my opinion the limitations are acceptable and we should not have the macro take strings because any path is still buildable with some tweaks like changing to bracket field access or converting strings)


### Macro Output:
- Currently, the macro just outputs AST literals, though of course it could be made to output anything, so either:
    - We implement Query for the AST nodes and just delete the existing types
    - Provide impls like `From<JPQueryAST> for JPQuery` for all existing structs then just have the macro call `.into()`
    - We have the macro output the current types(my least favorite, probably error prone and somewhat redundant)

### Testing:
- If using trybuild, we can either:
    - Create files dynamically from rfc test case file and then delete afterward to ensure accuracy and account for changes in the suite?
    - Just statically build the suite(I made a quick script), how often could it even change?(to me this feels like the answer)

thehiddenwaffle commented:
No problem, it's pretty huge

besok commented on jsonpath-ast/src/ast.rs:
```diff
@@ -0,0 +1,868 @@
+            fn from_pest(
+                pest: &mut ::from_pest::pest::iterators::Pairs<'pest, Rule>,
+            ) -> ::std::result::Result<Self, ::from_pest::ConversionError<::from_pest::Void>> {
+                let mut clone = pest.clone();
```
please better naming like `cloned_pest`

besok commented on jsonpath-ast/src/ast.rs:
```diff
@@ -0,0 +1,868 @@
+#[derive(Debug, new, PartialEq, FromPest)]
+#[cfg_attr(feature = "compiled-path", derive(Parse))]
+#[pest_ast(rule(Rule::main))]
+pub struct Main {
```
I would suggest better naming, like `JsPathMain` or `JsPathHead`, now when it appears in the other crates it is unclear what it is

----
ah, it is an intermdiaite state. The comment for the future final review then :)

besok commented on jsonpath-ast/src/ast.rs:
```diff
@@ -0,0 +1,868 @@
+    pub fn try_from_pest_parse(str: &str) -> Result<Self, ()> {
+        let mut rules = JSPathParser::parse(Rule::main, str).map_err(|_| ())?;
+        // *IF* the FromPest implementations are correctly written then Main::from_pest *cannot fail*
+        Main::from_pest(&mut rules).map_err(|_| ())
```
i would leave the return error or introduce a new one. Now, we loose the info even when we don't need, it is stil better to have the error

---
ah, it is an intermdiaite state. The comment for the future final review then :)

besok reviewed (commented):
Also, can you complement the description in the general readme with the changes:
  - what feature means
   - how to use it
  as soon as the pr is ready to merge.

besok commented:
First of all, I have to say, this is impressive work. Thank you.

thehiddenwaffle commented on src/parser/grammar/json_path_9535.pest:
```diff
@@ -1,6 +1,6 @@
+main = ${ SOI ~ jp_query ~ EOI }
 jp_query = {root ~ segments}
-segments = {(S ~ segment)*}
+segments = !{(S ~ segment)*}
```
I'm gonna quote the pest docs because I'm not the best at explaining and I'm relatively new to pest:
> Compound-atomic ($)
> Compound-atomic are identical to atomic rules with the exception that rules called by them are not forbidden from generating token pairs.
> 
> Non-atomic (!)
> 
> Non-atomic are identical to normal rules with the exception that they stop the cascading effect of atomic and compound-atomic rules.

So all the `not-atomic` does in this particular case is override the compound atomic for the segment rule and all downstream rules, allowing for everything except for the main rule to not be atomic. The main rule being atomic makes it so that leading and trailing white space on the main rule will be rejected by the parser, allowing us to [remove the validation logic in parser](https://github.com/besok/jsonpath-rust/pull/101/files/352becf9215a5676bdd4fcd60c14436e03ae66d9#diff-4a04259da480a6b794a2e947e4cc03eff4d1aa9330836f5b91cac68c5398193f) and allowing the FromPest impl to be more resilient. So adding the `!` to the child rules allows them to have leading/trailing white space again.

I guess I should have also added `!` to the root but in this case root does not parse to rules so it doesn't make a difference functionally.

thehiddenwaffle commented:
tests now take 4.5 sec to both build and run

thehiddenwaffle commented:
Added some final grammar improvements for more strict function name/arg checking, added preliminary docs. I think we can potentially close this one out as it's getting very long. If it's all good with you I'd like to follow it up with 2 more PR's, one to get full test coverage on the macro with trybuild, and a final one to actually provide `From<...AST> for ...` impls on the types in the main crate.

besok commented:
> It's up to you, I don't have a problem with adding docs as part of this PR, I'll just put it at the bottom for now and include something like " 🚧under construction🚧" or maybe "unstable", because with there PR there still won't be a use for all the stuff I built.

Ok, i will try to merge it this weekend then. Better to have smaller prs

besok reviewed (approved):
lgtm.
Thank you for the PR. 
Do you have plans of keep working on the feature further or we can publish a new version now? 
What do you think?

thehiddenwaffle commented:
Personally I think having a full test suite would probably best before a new version, I have already started on it and that the functions and JsonPath type system from the RFC are particularly tricky, perhaps next PR once I clean up my mess 🫢
